src/Storages/MergeTree/MutateTask.cpp

```diff
@@ -1338,12 +1346,18 @@ bool PartMergerWriter::mutateOriginalPartAndPrepareProjections()
     for (size_t i = 0, size = ctx->projections_to_build.size(); i < size; ++i)
     {
         const auto & projection = *ctx->projections_to_build[i];
-        auto & projection_squash = projection_squashes[i];
-        auto projection_block = projection_squash.add({});
-        if (projection_block)
+        auto & projection_squash_plan = projection_squash_plannings[i];
+        auto planned_chunk = projection_squash_plan.flush();
+        if (planned_chunk.hasChunkInfo())
         {
+            Chunk projection_chunk = projection_squashes[i].add(std::move(planned_chunk));
```

Comment:
You have several times situation when you need to combine plan + apply of squashing.
Lest make ApplySquashing as a pure function as far as it does not have a state between calls. It makes things easier.